### ISSUE TITLE
.NET Core Does Not work with SQL Named Pipes

### ISSUE THREAD

**Issue by dsartain18** [dotnet/runtime#41968]
### Description
I am on a laptop that has both SQL Server 2016 and SQL Server 2019 installed. I'm part of a development team where others don't necessarily have that same setup. I've gone through a lot of effort to make sure that our new solution of Function Apps, a Web App, integration tests, and an API can all load the database from the correct connection string setting for each given type of application. 

The problem is that for everyone else on my team, using "localhost" works just fine because they only have one version of SQL Server installed as the default version. For me, though, I have to work around this issue by setting up "localhost" as a named provider for the version of SQL Server I'm currently working with when working on longer projects. 

Historically that's been no problem because we were using .NET Framework. The new project is a mix of .NET Core and .NET Standard, and this is where my connection troubles began. 

I finally found this article explaining that .NET Core doesn't use Named Pipes because not every OS allows for Named Pipes: https://github.com/dotnet/runtime/issues/23032

For now, I have turned on TCP for SQL Server 2019 to get this working locally. However, I don't think this is the best solution. 

**Recommended Solution:**
Developers should be in control of whether to allow for something that may not exist in all cases. In my case, I am always coding against a Windows environment. So, I don't have to care what Linux doesn't support. 

Is there a way to set this up as an option to set during startup for whether or not to allow named pipes? Preferably an extension method that takes a bool parameter so it can be config driven per environment?

**Suggestion:**
Also, I researched and researched the "Named Pipes Can't Connect - 40" type error trying to fix a problem I knew I solved before to get other applications working. It wasn't until I included ".NET Core" in the search phrasing that I finally found an explanation that didn't point me right back to the same SQL Server Configuration settings I'd already double checked 50 times. 

Since we know this is a .NET Core issue, is there a way to explain that in the Exception message that is returned instead of leaving it the generic message that applies to .NET Framework?

**Comment by David-Engel** [comment]
@dsartain18 

> I finally found this article explaining that .NET Core doesn't use Named Pipes because not every OS allows for Named Pipes: #23032

I think you misunderstood what that issue said. .NET Core is the client in this situation. Your SQL Server instance may or may not be using named pipes to accept connections. That depends on how you have configured SQL Server, which is completely independent of .NET Core. .NET Core (specifically, the System.Data.SqlClient library) on Windows supports connecting to SQL Server via named pipes when the client is running on (normal) Windows. It does not support connecting to SQL Server via named pipes when the client is either a UWP application on Windows or running on Linux/macOS.

It's hard to tell from your issue exactly what your environment was when you encountered the error you mention. Are you building a UWP application? You also mentioned:

> I have to work around this issue by setting up "localhost" as a named provider for the version of SQL Server I'm currently working with

What exactly does that mean? Are you defining an alias for "localhost"? Are you installing named instances where you would connect via localhost\SQLServer2019 or localhost\SQLServer2017, for example?

**Comment by dsartain18** [comment]
Yes, I am using Named Instances on my box and setting up the “localhost” alias to point toward the instance I use the most often. 

I’m building a .NET Core 3.1 Function App and Web App. Neither project will allow “localhost” to work as a Named Pipe. I would point “localhost” to “\\ComputerName\SqlServerName” so all I have to reference in the config is “localhost” for the Server value.

I have gone into the SQL Configuration manager and enabled Named Pipes, and normally have TCP turned off outright to prevent any possibility of connection from outside sources when on shared public WiFi.

So, I’m not building UWP. I’m not running on Linux. But the localhost alias doesn’t seem to work over Named Pipes when using .NET Core but runs flawlessly under .NET Framework.

**Comment by cheenamalhotra** [comment]
Hi @dsartain18 

We're able to reproduce the issue as well.

Just to confirm we're on the same page, please also try to specify server name as `np:localhost` and test connection as that worked in our case. You can use that as a temporary workaround till we figure out the real problem.

We'll transfer this issue to sqlclient repository for further investigations.
